pipe.go

```diff
@@ -152,6 +152,14 @@ func _newPipe(connFn func() (net.Conn, error), option *ClientOption, r2ps bool)
 	p.pshks.Store(emptypshks)
 	p.clhks.Store(emptyclhks)
 
+	if option.AuthCredentialsFn != nil {
+		username, password, err := option.AuthCredentialsFn()
+		if err == nil {
+			option.Username = username
+			option.Password = password
```

Comment:
Hi @nirrattner,

Thank you for your contribution! I think this feature is really great. Just out of curiosity, are you building your own Redis service?

In terms of the implementation, however, we can't modify the `option.Username` and the `option.Password` field here, because the `option` is shared among all `newPipe` invocations. If we modify it, then we create races. We should introduce independent `username`, and `password` variables here.

Another question is why do you ignore the `err`? I thought that error was crucial for your authentication.

Reply:
Hi thanks for the feedback!

We are not building our own Redis service, but instead developing a Redis Module to support the alternate authentication mechanism within standard Redis.

I will incorporate your suggestion to avoid race conditions, thanks!

That is a good question about the choice to ignore the `err` -- my first thought was to allow the the client to fallback to any previously used credentials in the case of an `err`, and then depend on Redis to be the authority about whether these credentials are still valid. However, I'm rethinking this since it may be unintuitive behavior, especially because users will still have the option to implement any fallback logic within their `AuthCredentialsFn` implementation as needed, and there's likely an expectation that returning an error should fail the connection.